phpseclib/Crypt/Random.php

```diff
@@ -260,6 +260,44 @@ function crypt_random_string($length)
     }
 }
 
+if (!function_exists('phpseclib_safe_serialize')) {
+    /**
+     * Safely serialize variables
+     *
+     * If a class has a private __sleep() method it'll give a fatal error on PHP 5.2 and earlier.
+     * PHP 5.3 will emit a warning.
+     *
+     * @param mixed $arr
+     * @param array $refs optional
+     * @access public
+     */
+    function phpseclib_safe_serialize($arr, $refs = array())
+    {
+        if (is_object($arr)) {
+            return '';
+        }
+        if (!is_array($arr)) {
+            return serialize($arr);
+        }
+        $safearr = array();
+        foreach (array_keys($arr) as $key) {
+            if (is_object($arr[$key])) {
+                continue;
+            }
+            // without this recursive arrays (eg. $a = []; $a[] = &$a;) would result in an
+            // "Allowed memory size of ... bytes exhausted" Fatal error
+            foreach ($refs as $ref) {
+                if ($ref === $arr[$key]) {
```

Comment:
Yeah, I found that, too: It gave the following error on an object:
> Fatal error:  Uncaught Error: Cannot use object of type test as array in [...][...]:16

There seem to be a few issues with your update, though - I added comments to it.

I have an alternative approach: https://github.com/phpseclib/phpseclib/commit/062aa41ba002684101fe5dd526c161c9afa6085d. The change is:
* Instead of adding a `if (!is_object($arr))` before the recursion, it moves `if (isset($arr['__phpseclib_marker']))` out to the beginning of the recursion. The latter needs to be done **after** the former, instead of the same line, because it needs to already be sure that it's not an object at this point, because (from https://secure.php.net/manual/en/language.operators.precedence.php):

  > PHP does not (in the general case) specify in which order an expression is evaluated and code that assumes a specific order of evaluation should be avoided, because the behavior can change between versions of PHP or depending on the surrounding code.

Here is the ultimate test, which get passed on all versions (`private` and `$z` need to be commented out on lower versions):
```php
class test {
    private function __sleep() {
        return array();
    }
}

$aobj = new test();

$scalar_value = 'whatever';

$a['key1'] = 'value1';
$a[] = &$a;

$b[] = &$a;
$b[] = &$b;

$c[] = &$b;

$z = [];
$z[] = [&$z];

echo phpseclib_safe_serialize($GLOBALS);
```

What do you think?